src/group_theory/double_coset.lean

```diff
@@ -0,0 +1,256 @@
+/-
+Copyright (c) 2021 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import data.setoid.basic
+import group_theory.subgroup.basic
+import group_theory.coset
+import group_theory.subgroup.pointwise
+import data.set.basic
+import tactic.group
+
+/-!
+# Double cosets
+
+This file defines double cosets for two subgroups `H K` of a group `G` and the quotient of `G` by
+the double coset relation, i.e. `H \ G / K`. We also prove that `G` can be writen as a disjoint
+union of the double cosets and that if one of `H` or `K` is the trivial group (i.e. `⊥` ) then
+this is the usual left or right quotient of a group by a subgroup.
+
+## Main definitions
+
+* `double_coset_rel`: The double coset relation defined by two subgroups `H K` of `G`.
+* `double_coset.quotient`: The quotient of `G` by the double coset relation, i.e, ``H \ G / K`.
+-/
+
+
+variables {G : Type*} [group G] {α : Type*} [has_mul α] (J: subgroup G) (g : G)
+
+namespace double_coset
+
+open_locale pointwise
+
+/--The double_coset as an element of `set α` corresponding to `s a t` -/
+def doset (a : α) (s t : set α) : set α := s * {a} * t
+
+@[simp]
+lemma doset_mem {s t : set α} {a b : α} : b ∈ doset a s t ↔ ∃ (x ∈ s) (y ∈ t), b = x * a * y :=
+⟨λ ⟨_, y, ⟨x, _, hx, rfl, rfl⟩, hy, h⟩, ⟨x, hx, y, hy, h.symm⟩,
+  λ ⟨x, hx, y, hy, h⟩, ⟨x * a, y, ⟨x, a, hx, rfl, rfl⟩, hy, h.symm⟩⟩
+
+/--The relation given by saying two elements are related if their double cosets are equal -/
+def double_coset_rel (H K : set G) : G → G → Prop :=
+λ x y, doset x H K = doset y H K
+
+lemma rel_reflex {H K : set G} : reflexive (double_coset_rel H K) :=
+λ x, rfl
+
+lemma rel_symm {H K : set G} : symmetric (double_coset_rel H K) :=
+λ x y, eq.symm
+
+lemma rel_trans {H K : set G} : transitive (double_coset_rel H K) :=
+λ x y z, eq.trans
+
+lemma rel_is_equiv {H K : set G} : equivalence (double_coset_rel H K) :=
+⟨rel_reflex, rel_symm, rel_trans⟩
+
```

Comment:
```suggestion

end double_coset_rel

```

Reply:
I don't I had a `double_coset_rel` namespace, so are you suggesting I close the namespace `double_coset` here and have the rest in the root namespace?

Reply:
I don't think we should close any namespace here.